rfd/0112-automatic-database-users.md

```diff
@@ -0,0 +1,317 @@
+```
+
+Most commonly used databases (PostgreSQL, MySQL, MongoDB) support the concept
+of roles as collection of privileges within the database that can be assigned
+to a user so `db_roles` should extend well to other protocols we support too.
+
+### Conditions for creating users
+
+Teleport database service will attempt to auto-create the database user if
+both of these are true:
+
+- Teleport user's roleset includes a role that matches the database they're
+  connecting to (by `db_labels`) and database user they're connecting as (by
+  `db_users`) and has `create_db_user: true` option set.
+- The database they're connecting to has "adminUser" defined.
+
+### Creating users
+
+PostgreSQL users created by Teleport will be members of `teleport-auto-user`
+role. This is similar to Linux user provisioning which adds users to `teleport`
+Linux group for tracking/bookkeeping purposes.
+
+The `teleport-auto-user` role will be created by Teleport using "adminUser" and
+will not have any privileges:
+
+```sql
+create role teleport-auto-user
+```
+
+### Disabling users
+
+PostgreSQL users created by Teleport **will not be deleted** by Teleport.
+
+The main reason is dynamic users can have permissions to create database objects
+(depending on their database role) and users that own any database objects can't
+be deleted until the objects are deleted or their ownership is transfered to
+another database user.
+
+PostgreSQL provides commands for deleting objects owned by a user and changing
+their owner but it has to be run in each database where user has objects which
+would require some non-trivial database introspection. In addition, keeping
+original ownership preserves the audit trail.
+
+For these reasons, dynamic users instead will be **disabled instead of deleted**
+after each session: stripped of all roles except `teleport-auto-user` and
+updated with `nologin` trait so they can't connect.
+
+### Stored procedures
+
+To create and disable users Teleport will use 2 stored procedures:
+
+- `teleport_create_user(username varchar, roles varchar[])`
+- `teleport_disable_user(username varchar)`
+
+Usage of stored procedures is chosen for the following reasons:
+
+- Utility statements like create/grant do not support bound parameters so
+  using procedure is safer that building queries in the code. See security
+  section below for more details.
+- Stored procedures are "all or nothing" so if any of the roles are invalid
+  during user creation for example, the user won't be created and no extra
+  cleanup will be required.
+- Stored procedures make it easier to keep some of the control logic in the
+  database for avoid races between agents, for example only disabling the user
+  if it doesn't have any active sessions.
+
+The procedures' source code is included below.
+
+### Session flow
+
+When receiving a database client connection, Teleport database service will
+first connect to the database as the "adminUser", install both procedures
```

Comment:
> install both procedures described above

Can the procedures be installed when registering the database (and when catching Procedure-does-not-exist errors) to avoid extra calls every session? I guess we cannot audit log these if pre-install them.